### DISCUSSION TITLE
Setting initial routes with start- and endpoints

### DISCUSSION THREAD

**Discussion by FabianPieroth** [google/or-tools#2968] [Routing (and legacy CP) questions]
Thank you for your help and support. 

My issue is with giving the solver for my routing task initial routes as solution. Using the example from the website I could make it work for a simple case, however when I add start- and endpoints for the task I cannot get a feasible solution.

The function call routing.ReadAssignmentFromRoutes(data['initial_routes'], True) returns None all the time.

In the minimal example below I can print an initial solution when I use data['starts‘]=[1] as starting point. However, when using anything else, for example data['starts‘]=[2], it does not work at all.

As one leaves out the start and the endpoint in the given initial solution, it seems to me, that the starting point gets added to each of the indices given in data['initial_routes']. Giving negative indices does not work obviously, which makes the use of other starting points impossible.

```python
"""Vehicles Routing Problem (VRP)."""

from __future__ import print_function
from ortools.constraint_solver import pywrapcp
from ortools.constraint_solver import routing_enums_pb2



def create_data_model():
    """Stores the data for the problem."""
    data = {}
    data['distance_matrix'] = [
        [
            0, 0, 0, 0, 0
        ],
        [
            0, 1, 1, 1, 1
        ],
        [
            0, 1, 1, 1, 1
        ],
        [
            0, 1, 1, 1, 1
        ],
        [
            0, 1, 1, 1, 1
        ]
    ]
    data['initial_routes'] = [
        [3, 1, 2]
    ]
    data['num_vehicles'] = 1
    data['depot'] = 0
    data['starts'] = [1]
    data['ends'] = [0]
    return data


def print_solution(data, manager, routing, solution):
    """Prints solution on console."""
    max_route_distance = 0
    for vehicle_id in range(data['num_vehicles']):
        index = routing.Start(vehicle_id)
        plan_output = 'Route for vehicle {}:\n'.format(vehicle_id)
        route_distance = 0
        while not routing.IsEnd(index):
            plan_output += ' {} -> '.format(manager.IndexToNode(index))
            previous_index = index
            index = solution.Value(routing.NextVar(index))
            route_distance += routing.GetArcCostForVehicle(
                previous_index, index, vehicle_id)
        plan_output += '{}\n'.format(manager.IndexToNode(index))
        plan_output += 'Distance of the route: {}m\n'.format(route_distance)
        print(plan_output)
        max_route_distance = max(route_distance, max_route_distance)
    print('Maximum of the route distances: {}m'.format(max_route_distance))


def main():
    """Solve the CVRP problem."""
    # Instantiate the data problem.
    data = create_data_model()

    # Create the routing index manager.
    manager = pywrapcp.RoutingIndexManager(
        len(data['distance_matrix']),
        data['num_vehicles'],
        data['starts'],
        data['ends'])

    # Create Routing Model.
    routing = pywrapcp.RoutingModel(manager)

    # Create and register a transit callback.
    def distance_callback(from_index, to_index):
        """Returns the distance between the two nodes."""
        # Convert from routing variable Index to distance matrix NodeIndex.
        from_node = manager.IndexToNode(from_index)
        to_node = manager.IndexToNode(to_index)
        return data['distance_matrix'][from_node][to_node]

    transit_callback_index = routing.RegisterTransitCallback(distance_callback)

    # Define cost of each arc.
    routing.SetArcCostEvaluatorOfAllVehicles(transit_callback_index)

    # Add Distance constraint.
    dimension_name = 'Distance'
    routing.AddDimension(
        transit_callback_index,
        0,  # no slack
        3000,  # vehicle maximum travel distance
        True,  # start cumul to zero
        dimension_name)
    distance_dimension = routing.GetDimensionOrDie(dimension_name)
    distance_dimension.SetGlobalSpanCostCoefficient(100)

    initial_solution = routing.ReadAssignmentFromRoutes(data['initial_routes'], True)
    print('Initial solution:')
    print_solution(data, manager, routing, initial_solution)

    # Set default search parameters.
    search_parameters = pywrapcp.DefaultRoutingSearchParameters()
    search_parameters.first_solution_strategy = (
        routing_enums_pb2.FirstSolutionStrategy.PARALLEL_CHEAPEST_INSERTION
    )

    # Solve the problem.
    solution = routing.SolveWithParameters(search_parameters)

    # Print solution on console.
    if solution:
        print('Solution after search:')
        print_solution(data, manager, routing, solution)
        print('end')


if __name__ == '__main__':
    main()
```

**Comment by lperron** [comment]
I believe that if you change the depot, the initial route is no longer feasible.
Can you check this ?

**Comment by FabianPieroth** [comment]
Thank you for your reply!
So the term 
`data['depot']=0`
does not do anything here, as you can comment it out, sorry for leaving that in. The task here is just to find a route that passes through all nodes. 
If you meant setting the end point differently, you are correct. Setting
`data['ends']=[4]`
for example, the initial route is no longer feasible. Actually again I could not determine any feasible route without changing also data['starts'] to zero. Which again would speak in favour of my guess, that the starting point somehow gets added to the initial route.

**Comment by Mizux** [comment]
if you look at the sibling method https://github.com/google/or-tools/blob/39f44709bba203f5ff3bc18fab8098739f189a6d/ortools/constraint_solver/routing.h#L909-L911
you'll see:

> The indices of the outer vector in
  'routes' correspond to vehicles IDs, the inner vector contains the
  variable indices on the routes for the given vehicle. The inner vectors
  must not contain the start and end indices, as these are determined by the
  routing model.

Since your initial route is **`data['initial_routes'] = [[3, 1, 2]]`** then start and end node must be **`0`** or **`4`** that's why you had to changed `data['start'] = [0]` and `data['end'] = [4]`

@lperron maybe we should also add this comment to the `ReadAssignmentFromRoutes()` method too ?

**Comment by FabianPieroth** [comment]
Thank you again for your reply. 
That is exactly what I assumed how it should work. There are just some settings that leave me puzzled so far:

```
data['starts']=[2]
data['ends']=[4]
data['initial_routes']=[[0,1,3]]

Prints the route: 2 ->  0 ->  1 ->  3 -> 4
```

changing the endpoint to 3 gives here:

```
data['starts']=[2]
data['ends']=[3]
data['initial_routes']=[[0,1,3]]

Prints the route: 2 ->  0 ->  1 ->  4 -> 3
```

whereas when one changes the initial_route, swapping the 3 with a 4:

```
data['starts']=[2]
data['ends']=[3]
data['initial_routes']=[[0,1,4]]
```

gives an error, as 4 not being a valid index.

**Comment by tomliptrot** [comment]
@FabianPieroth did you ever get to the bottom of this one? I've got a similar problem.

**Comment by tomliptrot** [comment]
@Mizux 
I think it is something to do with this:

https://github.com/google/or-tools/blob/1da80508ba813ab44c1ebdedc1b37f301615df53/ortools/constraint_solver/routing.cc#L3415-L3419

As I get the error message 

> E0925 14:50:57.537616  3815 routing.cc:3328] Invalid index: 4

I guess that must be because `to_index >= Size()`. 

What is `Size()` here?

**Comment by tomliptrot** [comment]
Would removing `|| to_index >= Size()` solve the problem? Or would that cause problems elsewhere?

**Comment by tomliptrot** [comment]
My current work around is to order the distance matrix [starts, nodes, ends], which seems to solve the problem. But it's not ideal...

**Comment by FabianPieroth** [comment]
Hello @tomliptrot,
sorry for the late reply. Unfortunately I did not find a nice way to fix it. I guess I will try your approach of ordering the distance matrix. 
However, I have the feeling, that there is something easy that we are missing here.

**Comment by mertbakir** [comment]
Is there any progress or a better workaround on this? I'm having a similar ```invalid index``` issue. 

As far as I understand, this happens when the start end nodes are different and ending node is not the last node in the distance matrix. For example, in the above example:

This would work:
```
data['initial_routes'] = [[3, 1, 2]]
data['starts'] = [0]
data['ends'] = [4]
```

This will return ```invalid index: 4```
```
data['initial_routes'] = [[3, 1, 4]]
data['starts'] = [0]
data['ends'] = [2]
```

Starting node doesn't have to be the first one though. So, these work fine:
```
data['initial_routes'] = [[0, 1, 2]]
data['starts'] = [3]
data['ends'] = [4]
```

```
data['initial_routes'] = [[0, 3, 2]]
data['starts'] = [1]
data['ends'] = [4]
```

Is this because of the solver is actually solving a TSP, so it's designed the end at the starting location? The example in the website also starts and ends at the node 0. It would be better if we had an example for different start/end locations.

**Comment by mertbakir** [comment]
> My current work around is to order the distance matrix [starts, nodes, ends], which seems to solve the problem. But it's not ideal...

Hey @tomliptrot, have you found something better? This approach doesn't sound like it would work for multiple vehicles with different start/ends. How would you order the matrix?

**Comment by tomliptrot** [comment]
Hi, it was a while go, but I think what I did was just number keep all of the locations in the order `[[starts], [nodes], [ends]]`. 

You can have as many start locations as you want - eg `[0,1,2,3,4]` followed by nodes `[5,6...,n]` then ends locations `[n+1, n+2, ...]`

It feels a bit hacky but it seemed to work.

**Comment by mertbakir** [comment]
> Hi, it was a while go, but I think what I did was just number keep all of the locations in the order `[[starts], [nodes], [ends]]`.
> 
> You can have as many start locations as you want - eg `[0,1,2,3,4]` followed by nodes `[5,6...,n]` then ends locations `[n+1, n+2, ...]`
> 
> It feels a bit hacky but it seemed to work.

It seems that's what I'm going to end up if I can't find a better solution. Thanks.

**Comment by Mizux** [comment]
I think you need to use index (int64_t) not node.
Also as spotted by tomliprot RoutingIndexManager will allways move end node to end of list 

Please take a look at: https://github.com/google/or-tools/issues/2773#issuecomment-923765550
